miri-script/src/commands.rs

```diff
@@ -0,0 +1,652 @@
+        cmd!(sh, "rustup override set miri").run()?;
+        // Cleanup.
+        cmd!(sh, "cargo clean").run()?;
+        // Call `cargo metadata` on the sources in case that changes the lockfile
+        // (which fails under some setups when it is done from inside vscode).
+        let sysroot = cmd!(sh, "rustc +miri --print sysroot").read()?;
+        let sysroot = sysroot.trim();
+        let _ = cmd!(sh, "cargo metadata --format-version 1 --manifest-path {sysroot}/lib/rustlib/rustc-src/rust/compiler/rustc/Cargo.toml").ignore_stdout().run()?;
+        Ok(())
+    }
+
+    fn rustc_pull(&self, commit: Option<String>) -> Result<()> {
+        let sh = shell_with_parent_env()?;
+        sh.change_dir(&self.miri_dir);
+        let commit: String = commit.map(Result::Ok).unwrap_or_else(|| {
+            let rust_repo_head =
+                cmd!(sh, "git ls-remote https://github.com/rust-lang/rust/ HEAD").read()?;
+            rust_repo_head
+                .split_whitespace()
+                .next()
+                .map(|front| front.trim().to_owned())
+                .ok_or_else(|| anyhow!("Could not obtain Rust repo HEAD from remote."))
+        })?;
+        // Update rust-version file. As a separate commit, since making it part of
+        // the merge has confused the heck out of josh in the past.
+        sh.write_file(path!(self.miri_dir / "rust-version"), &commit)?;
+        const PREPARING_COMMIT_MESSAGE: &str = "Preparing for merge from rustc";
+        cmd!(sh, "git commit rust-version -m {PREPARING_COMMIT_MESSAGE}")
+            .run()
+            .context("FAILED to commit rust-version file, something went wrong")?;
+        // Fetch given rustc commit and note down which one that was
+        cmd!(sh, "git fetch http://localhost:8000/rust-lang/rust.git@{commit}{JOSH_FILTER}.git")
+            .run()
+            .context("FAILED to fetch new commits, something went wrong")?;
+        const MERGE_COMMIT_MESSAGE: &str = "Merge from rustc";
+        cmd!(sh, "git merge FETCH_HEAD --no-ff -m {MERGE_COMMIT_MESSAGE}")
+            .run()
+            .context("FAILED to merge new commits, something went wrong")?;
+        Ok(())
+    }
+
+    fn rustc_push(&self, github_user: &str, branch: &str) -> Result<()> {
+        let rustc_git = std::env::var_os("RUSTC_GIT").filter(|value| !value.is_empty());
+        let working_directory = rustc_git.unwrap_or_else(|| {
+            // If rustc_git is `Some`, we'll use an existing fork for the branch updates.
+            // Otherwise, do this in the local Miri repo.
+            println!("This will pull a copy of the rust-lang/rust history into this Miri checkout, growing it by about 1GB.");
+            println!("To avoid that, abort now and set the RUSTC_GIT environment variable to an existing rustc checkout. Proceed? [y/N] ");
+            let mut answer = String::new();
+            let _ = std::io::stdin().read_line(&mut answer);
```

Comment:
Why ignore errors here?

Reply:
Uh, I guess I did it out of laziness (not having to deal with type conversions to satisfy `unwrap_or_else` et al was nice). 
I've moved this snippet to `if let` now.